groups/bsl/bslstl/bslstl_optional.h

```diff
@@ -0,0 +1,1653 @@
+template <class TYPE>
+void Optional_DataImp<TYPE>::reset()
+BSLS_KEYWORD_NOEXCEPT
+{
+    if (d_hasValue) {
+        BloombergLP::bslma::DestructionUtil::destroy(&(d_buffer.object()));
+    }
+    d_hasValue = false;
+}
+
+//ACCESSORS
+template <class TYPE>
+inline
+bool
+Optional_DataImp<TYPE>::hasValue()
+    const BSLS_KEYWORD_NOEXCEPT
+{
+    if (d_hasValue) {
+        return true;                                                  // RETURN
+    }
+    return false;
+}
+
+template <class TYPE>
+inline
+TYPE&
+Optional_DataImp<TYPE>::value()
+    BSLS_KEYWORD_LVREF_QUAL
+{
+    BSLS_ASSERT(d_hasValue);
+
+    return d_buffer.object();
+}
+
+template <class TYPE>
+inline
+const TYPE&
+Optional_DataImp<TYPE>::value()
+    const BSLS_KEYWORD_LVREF_QUAL
+{
+    BSLS_ASSERT(d_hasValue);
+
+    return d_buffer.object();
+}
+
+#if defined(BSLS_COMPILERFEATURES_SUPPORT_REF_QUALIFIERS)
+template <class TYPE>
+inline
+TYPE&&
+Optional_DataImp<TYPE>::value()
+    BSLS_KEYWORD_RVREF_QUAL
+{
+    BSLS_ASSERT(d_hasValue);
+
+    return std::move(d_buffer.object());
+}
+
+template <class TYPE>
+inline
+const TYPE&&
+Optional_DataImp<TYPE>::value()
+    const BSLS_KEYWORD_RVREF_QUAL
+{
+    BSLS_ASSERT(d_hasValue);
+
+    return std::move(d_buffer.object());
+}
+#endif //defined(BSLS_COMPILERFEATURES_SUPPORT_REF_QUALIFIERS)
+
+                        // ===================
+                        // class Optional_Data
+                        // ===================
+
+template <class TYPE, bool IS_TRIVIALLY_DESTRUCTIBLE =
+                        BSLSTL_OPTIONAL_IS_TRIVIALLY_DESTRUCTIBLE<TYPE>::value>
+struct Optional_Data : public Optional_DataImp<TYPE>{
+    // Component-level class; DO NOT USE
+
+    // Optional_Data is a type used to manage 'value_type' object in
+    // 'optional'.  It allows 'optional<TYPE>' to be trivially destructible if
+    // 'TYPE' is trivially destructible.  This class also provides an
+    // abstraction for const value type.  An 'optional' may contain a 'const'
+    // type object.  An assignment to such an 'optional' should not succeed.
+    // However, unless the 'optional' itself is 'const', it should be possible
+    // to change the value of the 'optional' using 'emplace'.  In order to
+    // allow for that, this class manages a non-const object of 'value_type',
+    // but all the observers return a 'const' adjusted reference to the managed
+    // object.  This functionality is common for all value types and is
+    // implemented in the Optional_DataImp base class.  The derived class,
+    // 'Optional_Data', is specialised on 'value_type'
+    // 'is_trivially_destructible' trait.  The main template is for non
+    // trivially destructible types, and it provides a destructor that ensures
+    // the value_type destructor is called if 'd_buffer' holds an object.  The
+    // specialisation for 'is_trivially_destructible' types does not have a
+    // user provided destructor and 'is_trivially_destructible' itself.
```

Comment:
Please update this doc - refer to `Optional_DataImp` if necessary, but highlight that this is the base case for non-trivially-destructible types.

Reply:
Can you elaborate on that comment ? Which part needs to be updated ?